### PR TITLE
engine: propagate cancellation thru the watch loop

### PR DESCRIPTION
**Pull request by nicks**
Hello @maiamcc,

Please review the following commits I made in branch nicks/ch332/errors:

c9952782f8d835b6b13b8243e415da09c16522ef (2018-09-25 14:18:25 -0400)
engine: propagate cancellation thru the watch loop

### CODE REVIEW

File: internal/engine/build_and_deployer.go [hunk 1]
```diff
@@ -66,15 +67,28 @@ func (composite *CompositeBuildAndDeployer) BuildAndDeploy(ctx context.Context,
 	return BuildResult{}, lastErr
 }
 
+func isPermanentError(err error) bool {
+	if _, ok := err.(*model.ValidateErr); ok {
+		return true
+	}
+
+	cause := errors.Cause(err)
```
Comment (maiamcc) [inline internal/engine/build_and_deployer.go]: is this robust against fmt.errorf-ing on top of a context.Canceled error?

File: internal/engine/build_and_deployer.go [hunk 2]
```diff
@@ -66,15 +67,28 @@ func (composite *CompositeBuildAndDeployer) BuildAndDeploy(ctx context.Context,
 	return BuildResult{}, lastErr
 }
 
+func isPermanentError(err error) bool {
```
Comment (maiamcc) [inline internal/engine/build_and_deployer.go]: how are you defining "permanent error" here?

File: internal/engine/upper.go [hunk 3]
```diff
@@ -92,6 +92,8 @@ func (u Upper) CreateManifests(ctx context.Context, manifests []model.Manifest,
 		if err == nil {
 			buildStates[manifest.Name] = NewBuildState(buildResult)
 			lbs = append(lbs, k8s.ToLoadBalancerSpecs(buildResult.Entities)...)
+		} else if isPermanentError(err) {
```
Comment (maiamcc) [inline internal/engine/upper.go]: does this error get surfaced to the user in a useful way?

File: internal/engine/upper.go [hunk 4]
```diff
@@ -150,6 +152,9 @@ func (u Upper) CreateManifests(ctx context.Context, manifests []model.Manifest,
 					event.manifest,
 					buildState)
 				if err != nil {
+					if isPermanentError(err) {
```
Comment (maiamcc) [inline internal/engine/upper.go]: does this error get surfaced to the user in a useful way?

File: internal/engine/build_and_deployer.go (hunk 1 above)
Comment (nicks) [reply internal/engine/build_and_deployer.go]: no, that's why we use errors.Wrap! :raised_hands:

File: internal/engine/build_and_deployer.go (hunk 2 above)
Comment (nicks) [reply internal/engine/build_and_deployer.go]: added a comment!

File: internal/engine/upper.go (hunk 3 above)
Comment (nicks) [reply internal/engine/upper.go]: yes

File: internal/engine/upper.go (hunk 4 above)
Comment (nicks) [reply internal/engine/upper.go]: yes

**Comment by nicks** [comment]
screenshot from 2018-09-25 15-38-10

**Pull request merged**